libgalois/src/GraphTopology.cpp

```diff
@@ -435,6 +435,192 @@ katana::EdgeTypeAwareTopology::MakeFrom(
       pg, edge_type_index, e_topo, std::move(per_type_adj_indices)});
 }
 
+std::unique_ptr<katana::ProjectedShuffleTopology>
+katana::ProjectedShuffleTopology::Make(
+    const katana::PropertyGraph* pg, const std::vector<std::string>& node_types,
+    const std::vector<std::string>& edge_types) {
+  KATANA_LOG_DEBUG_ASSERT(pg);
+
+  const auto& topology = pg->topology();
+  if (topology.empty()) {
+    ProjectedShuffleTopology pst;
+    return std::make_unique<ProjectedShuffleTopology>(std::move(pst));
+  }
+
+  // calculate number of new nodes
+  uint32_t num_new_nodes = 0;
+  uint32_t num_new_edges = 0;
+
+  katana::DynamicBitset bitset_nodes;
+  bitset_nodes.resize(topology.num_nodes());
+
+  NUMAArray<Node> old_to_new_nodes_mapping;
+  old_to_new_nodes_mapping.allocateInterleaved(topology.num_nodes());
+
+  katana::ParallelSTL::fill(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      Node{0});
+
+  if (node_types.empty()) {
+    num_new_nodes = topology.num_nodes();
+    // set all nodes
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      bitset_nodes.set(src);
+      old_to_new_nodes_mapping[src] = 1;
+    });
+  } else {
+    std::set<katana::EntityTypeID> node_entity_type_ids;
+
+    for (auto node_type : node_types) {
+      auto entity_type_id = pg->GetNodeEntityTypeID(node_type);
+      node_entity_type_ids.insert(entity_type_id);
+    }
+
+    katana::GAccumulator<uint32_t> accum_num_new_nodes;
+    katana::GAccumulator<uint64_t> accum_num_new_edges;
+
+    katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+      for (auto type : node_entity_type_ids) {
+        if (pg->DoesNodeHaveType(src, type)) {
+          accum_num_new_nodes += 1;
+          accum_num_new_edges += std::distance(
+              topology.edges(src).begin(), topology.edges(src).end());
+          bitset_nodes.set(src);
+          old_to_new_nodes_mapping[src] = 1;
+          return;
+        }
+      }
+    });
+    num_new_nodes = accum_num_new_nodes.reduce();
+    num_new_edges = accum_num_new_edges.reduce();
+  }
+
+  // fill old to new nodes mapping
+  katana::ParallelSTL::partial_sum(
+      old_to_new_nodes_mapping.begin(), old_to_new_nodes_mapping.end(),
+      old_to_new_nodes_mapping.begin());
+
+  NUMAArray<Node> new_to_old_nodes_mapping;
+  new_to_old_nodes_mapping.allocateInterleaved(num_new_nodes);
+
+  katana::do_all(katana::iterate(topology.all_nodes()), [&](auto src) {
+    if (bitset_nodes.test(src)) {
+      old_to_new_nodes_mapping[src]--;
+      new_to_old_nodes_mapping[old_to_new_nodes_mapping[src]] = src;
```

Comment:
the index corresponding to src will only be updated in one thread, in which it is also used. I don't believe there is any concurrency issue here, since no two threads will either update/read the same location.